tests/framework/utils.py

```
@@ -56,61 +55,52 @@ def set_cpu_affinity(pid: int, cpulist: list) -> list:
-    try:
-        threads = psutil.Process(pid).threads()
-    except psutil.NoSuchProcess as exc:
-        logging.warning("Process %d does not exist", pid, exc_info=exc)
-        return {}
-
-    cpu_times = {}
-    for thread in threads:
-        try:
-            thread_name = get_thread_name(pid, thread.id)
-            cpu_times[thread_name] = CpuTimes(thread.user_time, thread.system_time)
-        except psutil.NoSuchProcess as exc:
-            logging.warning("Thread %d no longer exists", thread.id, exc_info=exc)
-            continue
-
-    return cpu_times
-
-
-def get_cpu_utilization(
-    pid: int,
-    interval: int = 1,
-    split_user_system: bool = False,
-) -> Dict[str, float | CpuTimes]:
-    """Return current process per thread CPU utilization over the interval (seconds)."""
-    cpu_utilization = {}
-    cpu_times_before = get_cpu_times(pid)
-    time.sleep(interval)
-    cpu_times_after = get_cpu_times(pid)
-    threads = set(cpu_times_before.keys()) & set(cpu_times_after.keys())
-    for thread_name in threads:
-        before = cpu_times_before[thread_name]
-        after = cpu_times_after[thread_name]
-        user = (after.user - before.user) / interval * 100
-        system = (after.system - before.system) / interval * 100
-        if split_user_system:
-            cpu_utilization[thread_name] = CpuTimes(user, system)
-        else:
-            cpu_utilization[thread_name] = user + system
-    return cpu_utilization
+class ProcessExt(psutil.Process):
+    """Extends psutil.Process with better thread support."""
+
+    def threads_cpu_times(self) -> Dict[str, CpuTimes]:
+        """Return a dict mapping thread name to CPU usage (in seconds) since start."""
+        cpu_times = {}
+        for pthread in super().threads():
+            try:
+                name = psutil.Process(pthread.id).name()
+                cpu_times[name] = CpuTimes(pthread.user_time, pthread.system_time)
+            except psutil.NoSuchProcess as exc:
```

maybe for these "nice if it works, don't care if it doesn't" kinda things, we can just use bare excepts? And maybe even just put the entire body of the function into one. That'd also save us the "except-pass" blocks at the call sites.

---

I wanted to let the caller know the process exited, to be consistent with the existing `psutil.Process` API. I am ignoring the error in that particular case to leave the behaviour as it used to be.

---

The reason for using the process and having the caller manage the process instance is to fix the issue of PID reuse. Then, I thought it'd be nice to just follow the same API as psutil, so that's why I'm extending the `psutil.Process` class to extend the support for threads. By extending the class, I'm keeping the same error handling as the original class (which basically just returns `NoSuchProcess` in normal conditions).
The issue of the error types was born from the fact I was reading procfs directly, and it's solved by using `psutil`, which converts those errors to `NoSuchProcess`.

If you don't like the class extension, I can revert back to the previous "util functions" pattern and return an empty dict, handling all errors internally.

```
def get_cpu_times(process: psutil.Process, ...):
    ...

def get_cpu_utilization(process: psutil.Process, ...):
    ...
```

---

now I'll just have to fight pylint to have a bare or broad except :D